unittests/TEST_pieces.cpp

```diff
@@ -0,0 +1,204 @@
+}
+
+TEST_F(PieceTests, TestRotation) {
+	Board board;
+	TBlock piece(board);
+
+	piece.positionRow = 0;
+	EXPECT_TRUE(piece.rotate()) << "should succeed, only obstructed by bottom boundary after rotation";
+	EXPECT_EQ(piece.rotation, 1) << "rotation state 0 -> 1";
+	EXPECT_EQ(piece.positionRow, 1) << "piece moves up to avoid bottom boundary 0 -> 1";
+
+	piece.positionCol = 0;
+	EXPECT_TRUE(piece.rotate()) << "should succeed, only obstructed by bottom boundary after rotation";
+	EXPECT_EQ(piece.rotation, 2) << "rotation state 1 -> 2";
+	EXPECT_EQ(piece.positionCol, 1) << "piece moves up to avoid left boundary 0 -> 1";
+
+	piece.rotate(); // rotation state is 3 now
+	piece.positionCol = 9;
+	EXPECT_TRUE(piece.rotate()) << "should succeed, only obstructed by bottom boundary after rotation";
+	EXPECT_EQ(piece.rotation, 0) << "rotation state 1 -> 2";
+	EXPECT_EQ(piece.positionCol, 8) << "piece moves up to avoid left boundary 0 -> 1";
+}
+
+TEST_F(PieceTests, TestRotationWithObstructions) {
+	Board board;
+	TBlock piece(board);
+
+	// make obstruction
+	board[{10, 4}] = 1;
+	board[{10, 5}] = 1;
+	board[{10, 6}] = 1;
+	
+	piece.positionRow = 11;
+	EXPECT_TRUE(piece.rotate()) << "piece collides with obstruction under, shoould succeed"; // rotation 1
+	EXPECT_EQ(piece.positionRow, 12) << "piece should move up to avoid collision with obstruction undernearth";
+	EXPECT_EQ(piece.positionCol, 5) << "col should not change";
+
+	// more obstructions
+	for (int r = 0, re = 7; r < re; ++r) {
+		board[{r, 2}] = 1;
+		board[{r, 7}] = 1;
+		board[{r, 3}] = 1;
+		board[{r, 6}] = 1;
+	}
+	
+	piece.positionRow = 6;
+	piece.positionCol = 0;
+	EXPECT_TRUE(piece.rotate()) << "rotation should succeed";// rotation 2
+	EXPECT_FALSE(piece.checkCollidesAtOffset(0, 0)) << "piece should not collide with any obstruction or boundary";
+	EXPECT_EQ(piece.positionCol, 1) << "rotates along left boundary, needs to move 1 left";
+	EXPECT_EQ(piece.positionRow, 7) << "rotates along left boundary, needs to move 1 up";
+
+	piece.rotate(); // rotation 3
+	piece.positionRow = 6;
+	piece.positionCol = 9;
+	EXPECT_TRUE(piece.rotate()) << "rotation should succeed"; // rotation 0
+	EXPECT_FALSE(piece.checkCollidesAtOffset(0, 0)) << "piece should not collide with any obstruction or boundary";
+	EXPECT_EQ(piece.positionCol, 8)<< "rotates along right boundary, needs to move right";
+	EXPECT_EQ(piece.positionRow, 7)<< "rotates along right boundary, needs to move 1 up";
+
+	piece.rotate(); // rotation 1
+	piece.positionRow = 5;
+	piece.positionCol = 4;
+	EXPECT_TRUE(piece.rotate()) << "rotation should succeed"; // rotation 2
+	EXPECT_FALSE(piece.checkCollidesAtOffset(0, 0)) << "piece should not collide with any obstruction or boundary";
+	EXPECT_EQ(piece.positionCol, 4) << "rotates in tight space between set bocks, does not move horizontally";
+	EXPECT_EQ(piece.positionRow, 7) << "rotates in tight space between set bocks, needs to move 2 up";
```

Comment:
"rotates in tight space between set bocks, does not move horizontally" is a better assertion failure message than "rotates in tight space between set bocks, needs to move 2 up"